### DISCUSSION TITLE
(TypeScript) Export Loadable type

### DISCUSSION THREAD

**Discussion by edongashi** [pmndrs/jotai#1532] [Ideas] [answered]
When I want to write utility hooks like the following:

```ts
const myDataAtom = atom(...);
const myDataAtomLoadable = loadable(myAtom);

export function useMyData(): MyData {
  return useAtomValue(myDataAtom);
}

export function useMyDataLoadable(): Loadable<MyData> /* <--- not exposed */ { 
  return useAtomValue(myDataAtomLoadable);
}
```

I could let type inference do the job but we have a linter rule to be explicit about return types of `export`ed functions.
I had to copy and paste `Loadable<Value>` declaration into the codebase.

Another type that I found missing was `WriteGetter` which I needed to abstract over some atom write action.
I'm using `Getter` for the time being because this one is much more complex to copy over.

**Comment by dai-shi** [comment, marked as answer]
Related: #1467 

> but we have a linter rule to be explicit about return types of exported functions.

Ah, I'm aware of that rule... but not super happy with it.

Let's seek what would be possible with that rule.

For WriteGetter, does this work?
```ts
type WriteGetter = Parameters<WritableAtom<unknown, unknown>['write']>[0]
```

I think Loadable needs to be copied, because we can't extract types with generics. Coping does actually make sense for the lint rule, unless it's very complicated.

**Reply by dai-shi** [reply]
That said, exporting Loadable is something that can be considered.

**Reply by edongashi** [reply]
Thank you for the quick response!

> Related: https://github.com/pmndrs/jotai/issues/1467

Oops, sorry! I tried to search for related but it must have slipped.

> For WriteGetter, does this work?

That's smart and seems to work, thanks!

> I think Loadable needs to be copied, because we can't extract types with generics. Coping does actually make sense for the lint rule, unless it's very complicated.

Haven't been aware of the limitations. In any case, I'm fine with copying small snippets or typing some simplified version, and it helps that you agree with the approach.

**Reply by edongashi** [reply]
> That said, exporting Loadable is something that can be considered.

Renaming the discussion to "Export Loadable type" to narrow it down and not overlap with the previous discussion.

**Comment by dai-shi** [comment]
Looking at your example again, I don't think you need generics.

```ts
export function useMyDataLoadable(): ExtractAtomValue<typeof myDataAtomLoadable>
  return useAtomValue(myDataAtomLoadable);
}
```
should work?

**Reply by edongashi** [reply]
Typescript doesn't complain so it seems to work. Too opaque for my taste though and `Loadable` is only 4 lines.

However, these are really good examples on how one can extract hidden types.

**Reply by dai-shi** [reply]
Declaring explicit Loadable type (i.e. copy&paste) means you "expect" concrete type, regardless of how the library internal type is.

`ExtractAtomValue<typeof myDataAtomLoadable>` is like passing through the type to just avoid the lint rule.

Both should work and they mean different things. For now, I think they are valid solutions. I'm still open for comments from others.
